### PR TITLE
Add symbols file

### PR DESCRIPTION
**Pull request by agx**
to only export symbols matching wlr_* from the wlroots library

### CODE REVIEW

**Comment by ascent12** [comment]
We still use the wlr_ prefix for a lot of our internal stuff, so I would prefer that they're listed out explicitly.
To save you some effort, here is the list of functions generated by
`find include/wlr/ -type f | xargs ctags -x --c-kinds=p | cut -d' ' -f1`
```
_strip_path
egl_error
usec_to_msec
wlr_backend_autocreate
wlr_backend_destroy
wlr_backend_get_egl
wlr_backend_get_renderer
wlr_backend_init
wlr_backend_is_drm
wlr_backend_is_headless
wlr_backend_is_libinput
wlr_backend_is_multi
wlr_backend_is_wl
wlr_backend_is_x11
wlr_backend_start
wlr_box_closest_point
wlr_box_contains_point
wlr_box_empty
wlr_box_intersection
wlr_box_rotated_bounds
wlr_box_transform
wlr_compositor_create
wlr_compositor_destroy
wlr_cursor_absolute_to_layout_coords
wlr_cursor_attach_input_device
wlr_cursor_attach_output_layout
wlr_cursor_create
wlr_cursor_destroy
wlr_cursor_detach_input_device
wlr_cursor_map_input_to_output
wlr_cursor_map_input_to_region
wlr_cursor_map_to_output
wlr_cursor_map_to_region
wlr_cursor_move
wlr_cursor_set_image
wlr_cursor_set_surface
wlr_cursor_warp
wlr_cursor_warp_absolute
wlr_data_device_manager_create
wlr_data_device_manager_destroy
wlr_data_source_finish
wlr_data_source_init
wlr_drm_backend_create
wlr_egl_bind_display
wlr_egl_create_image
wlr_egl_create_surface
wlr_egl_destroy_image
wlr_egl_finish
wlr_egl_init
wlr_egl_make_current
wlr_egl_query_buffer
wlr_egl_swap_buffers
wlr_gamma_control_manager_create
wlr_gamma_control_manager_destroy
wlr_gles2_renderer_create
wlr_headless_add_input_device
wlr_headless_add_output
wlr_headless_backend_create
wlr_idle_create
wlr_idle_destroy
wlr_idle_notify_activity
wlr_input_device_destroy
wlr_input_device_init
wlr_input_device_is_headless
wlr_input_device_is_libinput
wlr_input_device_is_wl
wlr_input_device_is_x11
wlr_keyboard_destroy
wlr_keyboard_get_modifiers
wlr_keyboard_init
wlr_keyboard_led_update
wlr_keyboard_notify_key
wlr_keyboard_notify_modifiers
wlr_keyboard_set_keymap
wlr_keyboard_set_repeat_info
wlr_libinput_backend_create
wlr_libinput_get_device_handle
wlr_list_cat
wlr_list_del
wlr_list_find
wlr_list_finish
wlr_list_for_each
wlr_list_init
wlr_list_insert
wlr_list_peek
wlr_list_pop
wlr_list_push
wlr_list_qsort
wlr_log_init
wlr_matrix_identity
wlr_matrix_mul
wlr_matrix_project_box
wlr_matrix_rotate
wlr_matrix_scale
wlr_matrix_texture
wlr_matrix_transform
wlr_matrix_translate
wlr_multi_backend_add
wlr_multi_backend_create
wlr_multi_backend_remove
wlr_multi_get_session
wlr_multi_is_empty
wlr_output_create_global
wlr_output_cursor_create
wlr_output_cursor_destroy
wlr_output_cursor_move
wlr_output_cursor_set_image
wlr_output_cursor_set_surface
wlr_output_damage_add
wlr_output_damage_add_box
wlr_output_damage_add_whole
wlr_output_damage_create
wlr_output_damage_destroy
wlr_output_damage_make_current
wlr_output_damage_swap_buffers
wlr_output_destroy
wlr_output_destroy_global
wlr_output_effective_resolution
wlr_output_enable
wlr_output_from_resource
wlr_output_get_gamma_size
wlr_output_init
wlr_output_is_drm
wlr_output_is_headless
wlr_output_is_wl
wlr_output_is_x11
wlr_output_layout_add
wlr_output_layout_add_auto
wlr_output_layout_closest_point
wlr_output_layout_contains_point
wlr_output_layout_create
wlr_output_layout_destroy
wlr_output_layout_get
wlr_output_layout_get_box
wlr_output_layout_get_center_output
wlr_output_layout_intersects
wlr_output_layout_move
wlr_output_layout_output_at
wlr_output_layout_output_coords
wlr_output_layout_remove
wlr_output_make_current
wlr_output_schedule_frame
wlr_output_send_frame
wlr_output_set_custom_mode
wlr_output_set_fullscreen_surface
wlr_output_set_gamma
wlr_output_set_mode
wlr_output_set_position
wlr_output_set_scale
wlr_output_set_transform
wlr_output_swap_buffers
wlr_output_transform_compose
wlr_output_transform_invert
wlr_output_transformed_resolution
wlr_output_update_custom_mode
wlr_output_update_enabled
wlr_output_update_mode
wlr_output_update_needs_swap
wlr_pointer_destroy
wlr_pointer_init
wlr_primary_selection_device_manager_create
wlr_primary_selection_device_manager_destroy
wlr_primary_selection_source_finish
wlr_primary_selection_source_init
wlr_region_create
wlr_region_expand
wlr_region_from_resource
wlr_region_scale
wlr_region_transform
wlr_render_colored_ellipse
wlr_render_colored_quad
wlr_render_texture_create
wlr_render_with_matrix
wlr_renderer_begin
wlr_renderer_buffer_is_drm
wlr_renderer_clear
wlr_renderer_destroy
wlr_renderer_end
wlr_renderer_format_supported
wlr_renderer_get_formats
wlr_renderer_init
wlr_renderer_read_pixels
wlr_renderer_scissor
wlr_screenshooter_create
wlr_screenshooter_destroy
wlr_seat_client_for_wl_client
wlr_seat_client_from_resource
wlr_seat_client_send_primary_selection
wlr_seat_client_send_selection
wlr_seat_create
wlr_seat_destroy
wlr_seat_get_keyboard
wlr_seat_keyboard_clear_focus
wlr_seat_keyboard_end_grab
wlr_seat_keyboard_enter
wlr_seat_keyboard_has_grab
wlr_seat_keyboard_notify_enter
wlr_seat_keyboard_notify_key
wlr_seat_keyboard_notify_modifiers
wlr_seat_keyboard_send_key
wlr_seat_keyboard_send_modifiers
wlr_seat_keyboard_start_grab
wlr_seat_pointer_clear_focus
wlr_seat_pointer_end_grab
wlr_seat_pointer_enter
wlr_seat_pointer_has_grab
wlr_seat_pointer_notify_axis
wlr_seat_pointer_notify_button
wlr_seat_pointer_notify_enter
wlr_seat_pointer_notify_motion
wlr_seat_pointer_send_axis
wlr_seat_pointer_send_button
wlr_seat_pointer_send_motion
wlr_seat_pointer_start_grab
wlr_seat_pointer_surface_has_focus
wlr_seat_set_capabilities
wlr_seat_set_keyboard
wlr_seat_set_name
wlr_seat_set_primary_selection
wlr_seat_set_selection
wlr_seat_touch_end_grab
wlr_seat_touch_get_point
wlr_seat_touch_has_grab
wlr_seat_touch_notify_down
wlr_seat_touch_notify_motion
wlr_seat_touch_notify_up
wlr_seat_touch_num_points
wlr_seat_touch_point_clear_focus
wlr_seat_touch_point_focus
wlr_seat_touch_send_down
wlr_seat_touch_send_motion
wlr_seat_touch_send_up
wlr_seat_touch_start_grab
wlr_seat_validate_grab_serial
wlr_server_decoration_manager_create
wlr_server_decoration_manager_destroy
wlr_server_decoration_manager_set_default_mode
wlr_session_change_vt
wlr_session_close_file
wlr_session_create
wlr_session_destroy
wlr_session_find_gpus
wlr_session_open_file
wlr_session_signal_add
wlr_surface_create
wlr_surface_from_resource
wlr_surface_get_main_surface
wlr_surface_get_matrix
wlr_surface_has_buffer
wlr_surface_make_subsurface
wlr_surface_send_enter
wlr_surface_send_frame_done
wlr_surface_send_leave
wlr_surface_set_role
wlr_surface_set_role_committed
wlr_surface_subsurface_at
wlr_tablet_pad_destroy
wlr_tablet_pad_init
wlr_tablet_tool_destroy
wlr_tablet_tool_init
wlr_texture_bind
wlr_texture_destroy
wlr_texture_get_buffer_size
wlr_texture_get_matrix
wlr_texture_init
wlr_texture_update_pixels
wlr_texture_update_shm
wlr_texture_upload_drm
wlr_texture_upload_eglimage
wlr_texture_upload_pixels
wlr_texture_upload_shm
wlr_touch_destroy
wlr_touch_init
wlr_wl_backend_create
wlr_wl_output_create
wlr_wl_shell_create
wlr_wl_shell_destroy
wlr_wl_shell_surface_configure
wlr_wl_shell_surface_ping
wlr_wl_shell_surface_popup_at
wlr_x11_backend_create
wlr_xcursor_frame
wlr_xcursor_get_resize_name
wlr_xcursor_manager_create
wlr_xcursor_manager_destroy
wlr_xcursor_manager_get_xcursor
wlr_xcursor_manager_load
wlr_xcursor_manager_set_cursor_image
wlr_xcursor_theme_destroy
wlr_xcursor_theme_get_cursor
wlr_xcursor_theme_load
wlr_xdg_shell_create
wlr_xdg_shell_destroy
wlr_xdg_shell_v6_create
wlr_xdg_shell_v6_destroy
wlr_xdg_surface_ping
wlr_xdg_surface_popup_at
wlr_xdg_surface_popup_get_position
wlr_xdg_surface_v6_ping
wlr_xdg_surface_v6_popup_at
wlr_xdg_surface_v6_popup_get_position
wlr_xdg_toplevel_send_close
wlr_xdg_toplevel_set_activated
wlr_xdg_toplevel_set_fullscreen
wlr_xdg_toplevel_set_maximized
wlr_xdg_toplevel_set_resizing
wlr_xdg_toplevel_set_size
wlr_xdg_toplevel_v6_send_close
wlr_xdg_toplevel_v6_set_activated
wlr_xdg_toplevel_v6_set_fullscreen
wlr_xdg_toplevel_v6_set_maximized
wlr_xdg_toplevel_v6_set_resizing
wlr_xdg_toplevel_v6_set_size
wlr_xwayland_create
wlr_xwayland_destroy
wlr_xwayland_set_cursor
wlr_xwayland_set_seat
wlr_xwayland_surface_activate
wlr_xwayland_surface_close
wlr_xwayland_surface_configure
wlr_xwayland_surface_set_fullscreen
wlr_xwayland_surface_set_maximized
xwm_create
xwm_destroy
xwm_handle_selection_event
xwm_selection_finish
xwm_selection_init
xwm_set_cursor
xwm_set_seat
```

**Comment by emersion** [comment]
Can't we automatically generate these with the command you posted? Keeping an up-to-date list will be a pain.

**Comment by agx** [comment]
@emersion wouldn't it make sense to move everything exported (that is intended to be used by compositors) under the wlr_* namespace and have everything else outside of it?  That's the list of symbols from the symbols file compared to the list generated by @ascent12 above:

```
+os_create_anonymous_file
-usec_to_msec
+wlr_data_device_keyboard_drag_interface
+wlr_data_device_pointer_drag_interface
+wlr_data_device_touch_drag_interface
+wlr_drm_backend_get_session
+wlr_drm_check_features
+wlr_drm_connector_cleanup
+wlr_drm_connector_start_renderer
+wlr_drm_event
+wlr_drm_get_connector_props
+wlr_drm_get_crtc_props
+wlr_drm_get_plane_props
+wlr_drm_get_prop
+wlr_drm_get_prop_blob
+wlr_drm_plane_surfaces_init
+wlr_drm_renderer_finish
+wlr_drm_renderer_init
+wlr_drm_resources_free
+wlr_drm_resources_init
+wlr_drm_restore_outputs
+wlr_drm_scan_connectors
+wlr_drm_surface_finish
+wlr_drm_surface_get_front
+wlr_drm_surface_init
+wlr_drm_surface_make_current
+wlr_drm_surface_mgpu_copy
+wlr_drm_surface_post
+wlr_drm_surface_swap_buffers
+wlr_egl_get_buffer_age
+wlr_libinput_event
+wlr_libinput_keyboard_create
+wlr_libinput_pointer_create
+wlr_libinput_tablet_pad_create
+wlr_libinput_tablet_tool_create
+wlr_libinput_touch_create
+_wlr_log
+wlr_log_stderr
+WLROOTS_0_0_0 A
+wlr_output_layout_adjacent_output
+wlr_signal_emit_safe
+wlr_subsurface_destroy
+wlr_wl_output_for_surface
+wlr_wl_output_layout_get_box
+wlr_wl_output_move_cursor
+wlr_wl_output_update_cursor
+wlr_wl_registry_poll
-xwm_create
-xwm_destroy
-xwm_handle_selection_event
-xwm_selection_finish
-xwm_selection_init
-xwm_set_cursor
-xwm_set_seat
```

**Comment by agx** [comment]
I've updated the PR to exclude wlr_* symbols that probably shouldn't be exported.

File: wlroots.syms [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        wlr_egl_get_buffer_age;
+        wlr_libinput_event;
+        wlr_libinput_keyboard_create;
+        wlr_libinput_pointer_create;
+        wlr_libinput_tablet_pad_create;
+        wlr_libinput_tablet_tool_create;
+        wlr_libinput_touch_create;
+        _wlr_log;
+        wlr_log_stderr;
+        wlr_output_layout_adjacent_output;
```
Comment (emersion) [inline wlroots.syms]: This one should be exported, actually.

**Comment by emersion** [comment]
>wouldn't it make sense to move everything exported (that is intended to be used by compositors) under the wlr_* namespace and have everything else outside of it?

Hum, that makes sense. Trying to think about edge cases.

**Comment by Ongy** [comment]
The edge case of debugging.
When I attach a debugger, get stack traces or try to debug other issues, I'd appreciate unique symbol names.

File: wlroots.syms [hunk 2]
```diff
@@ -0,0 +1,53 @@
+WLROOTS_0_0_0 {
+    global:
+        wlr_*;
+        # FIXME: these shouldn't be exported or renamed
+        os_create_anonymous_file;
+        _strip_path;
+        _wlr_log;
```
Comment (SirCmpwn) [inline wlroots.syms]: These guys should both be exported

File: wlroots.syms [hunk 3]
```diff
@@ -0,0 +1,53 @@
+WLROOTS_0_0_0 {
+    global:
+        wlr_*;
+        # FIXME: these shouldn't be exported or renamed
+        os_create_anonymous_file;
```
Comment (SirCmpwn) [inline wlroots.syms]: This one shouldn't be though

**Comment by SirCmpwn** [comment]
Would prefer to see this fix the FIXMEs rather than just document them and leave it for later

File: wlroots.syms (hunk 3 above)
Comment (agx) [reply wlroots.syms]: Wouldn't this be useful as wlr_create_anonyous_file ?

Comment (emersion) [reply wlroots.syms]: This is exported only because it's used by the screenshooter client. I don't think we want to expose it for compositors.

Comment (emersion) [reply wlroots.syms]: Maybe it would be better to duplicate this function in the screenshooter instead of exporting it.

**Comment by ascent12** [comment]
I really don't think we need the former `os_create_anonymous` in our API. It's not core to what wlroots is doing at all.
I don't like the idea of exposing extraneous util code which has nothing to do with writing a wayland compositor.

I think it just needs to be removed from examples/screenshot.c. Its use in there is not even that important; it could easily be replaced by a simple call to mkstemp.

**Comment by agx** [comment]
@Ongy but static functions already lack the wlr_ prefix and from what I can see the ones that we export accidentally atm just have a wlr_prefix because they're shared between internal code? 
So what about keeping the wlr_ prefix and just listing them explicitly (like in the current version)? 
Another possibility would be to switch to a _wlr_ or wlri_ prefix?

**Comment by Ongy** [comment]
Static functions don't usually get symbols.
Which is fine, since that's one of the features of static, but it does annoy me when I want to debug things from time to time...

I don't care about keeping the `wlr_` prefix for them. If they end up being `wlri_` or something similar that's fine by me.
I just wouldn't want them to be generic names, that may collide with other projects. Even if we technically could use them.

**Comment by SirCmpwn** [comment]
>wlri_

-1

File: examples/screenshot.c [hunk 4]
```diff
@@ -121,12 +121,33 @@ static const struct wl_registry_listener registry_listener = {
 	.global_remove = handle_global_remove,
 };
 
+
+int backingfile(off_t size) {
```
Comment (emersion) [inline examples/screenshot.c]: This should be `static`

**Comment by agx** [comment]
Updated.

**Comment by emersion** [comment]
Thanks!

**Pull request merged**